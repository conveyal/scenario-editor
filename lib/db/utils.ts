import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollectionResults(obj: unknown) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollectionResults)
  if (typeof obj === 'object') return mapValues(obj, serializeCollectionResults)
  return obj
}
