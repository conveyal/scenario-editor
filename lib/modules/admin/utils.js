import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect(Component) {
  return connect(createStructuredSelector(select), actions)(Component)
}

const add0 = n => (n >= 10 ? n : `0${n}`)

export function msToDuration(ms) {
  const durationS = Math.floor((Date.now() - ms) / 1000)
  const s = durationS % 60
  const m = Math.floor((durationS / 60) % 60)
  const h = Math.floor((durationS / 60 / 60) % 60)

  return `${add0(h)}:${add0(m)}:${add0(s)}`
}

/**
 * Example commit string: "v3.4.0-118-gbe7199c". Take the last section after
 * splitting on the hyphens and remove the "g".
 */
const githubUrl = 'https://github.com/conveyal/r5/commits'
export function createR5Url(workerCommit) {
  // not a release version
  if (workerCommit.indexOf('-') !== -1) {
    const hash = workerCommit.split('-')[2].slice(1) //
    return `${githubUrl}/${hash}`
  }
  return `${githubUrl}/${workerCommit}`
}

export function createWorkerUrl(instanceId, region) {
  return `https://${region}.console.aws.amazon.com/ec2/v2/home?region=${region}#Instances:instanceId=${instanceId};sort=tag:Name`
}

const C1 = '#2389c9'
const C2 = '#36c923'
const C3 = '#aab4bb'

/**
 * Create a progress bar with a linear-gradient.
 */
export function createLinearGradientForJob(c, d, t) {
  const p1 = (c / t) * 100
  const p2 = (d / t) * 100
  return `linear-gradient(to right,
    ${C1} 0%,
    ${C1} ${p1}%,
    ${C2} ${p1}%,
    ${C2} ${p2}%,
    ${C3} ${p2}%,
    ${C3} 100%)`
}
