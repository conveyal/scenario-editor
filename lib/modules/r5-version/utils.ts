import {
  MINIMUM_R5_VERSION,
  VERSION_PARSE_REGEX,
  OLD_VERSION_PARSE_REGEX
} from './constants'

/**
 * Convert an r5 version to a number for comparing
 * NB Assumes no major, minor, patch, or commit version exceeds 100
 */
export function versionToNumber(version?: string): number {
  const versionString = version || ''
  if (versionString.match(/\./g) && versionString.match(/\./g).length == 1) {
    const versionMatch = versionString.match(VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [, major = '0', minor = '0', commit = '0'] = versionMatch

      return (
        parseInt(major) * 10000 + parseInt(minor) * 100 + parseInt(commit) / 100
      )
    }

    return 0
  } else {
    const versionMatch = versionString.match(OLD_VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [
        ,
        major = '0',
        minor = '0',
        patch = '0',
        commit = '0'
      ] = versionMatch

      return (
        parseInt(major) * 10000 +
        parseInt(minor) * 100 +
        parseInt(patch) +
        parseInt(commit) / 100
      )
    }

    return 0
  }
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = (version || '').search(/-/g) !== -1
  return isTestVersion || workerVersionInRange(version, min, max)
}
