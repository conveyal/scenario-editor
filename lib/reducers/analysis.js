
export const reducers = {
  'set profile request' (state, action) {
    return {
      ...state,
      profileRequest: action.payload
    }
  },
  'set destination grid' (state, action) {
    return {
      ...state,
      destinationGrid: action.payload
    }
  },
  'set destination' (state, action) {
    return {
      ...state,
      destination: action.payload
    }
  },
  'clear isochrone results' (state, action) {
    return {
      ...state,
      isochrone: null,
      isochroneLonLat: null
    }
  },
  'set isochrone cutoff' (state, action) {
    return {
      ...state,
      isochroneCutoff: action.payload
    }
  },
  'set isochrone fetch status' (state, action) {
    return {
      ...state,
      isFetchingIsochrone: action.payload
    }
  },
  'set isochrone fetch status message' (state, action) {
    return {
      ...state,
      isochroneFetchStatusMessage: action.payload
    }
  },
  'set isochrone lonlat' (state, action) {
    return {
      ...state,
      isochroneLonLat: action.payload
    }
  },
  'set scenario application errors' (state, action) {
    return {
      ...state,
      scenarioApplicationErrors: action.payload
    }
  },
  'set scenario application warnings' (state, action) {
    return {
      ...state,
      scenarioApplicationWarnings: action.payload
    }
  },
  'set travel time surface' (state, action) {
    return {
      ...state,
      travelTimeSurface: action.payload
    }
  },
  'set comparison travel time surface' (state, action) {
    return {
      ...state,
      comparisonTravelTimeSurface: action.payload
    }
  },
  'enter analysis mode' (state, action) {
    return {
      ...state,
      active: true
    }
  },
  'exit analysis mode' (state, action) {
    return {
      ...state,
      scenarioApplicationErrors: null,
      scenarioApplicationWarnings: null,
      travelTimeSurface: null,
      comparisonTravelTimeSurface: null,
      active: false
    }
  },
  'set active variant' (state, action) {
    return {
      ...state,
      activeVariant: action.payload
    }
  },
  'clear comparison scenario' (state) {
    return {
      ...state,
      comparisonScenarioId: null,
      comparisonBundleId: null,
      comparisonVariant: null,
      comparisonModifications: null
    }
  },
  'set comparison scenario' (state, action) {
    const { id, bundleId, variantIndex } = action.payload
    return {
      ...state,
      comparisonScenarioId: id,
      comparisonBundleId: bundleId,
      comparisonVariant: variantIndex,
      comparisonModifications: null
    }
  },
  'set comparison modifications' (state, action) {
    // re-set all these to ensure the ID and the modifications cannot get out of sync
    const { id, bundleId, variantIndex, modifications } = action.payload
    return {
      ...state,
      comparisonScenarioId: id,
      comparisonBundleId: bundleId,
      comparisonVariant: variantIndex,
      comparisonModifications: modifications
    }
  },
  'set active regional analyses' (state, action) {
    const { id, comparisonId } = action.payload
    return {
      ...state,
      regional: {
        id,
        comparisonId,
        // clear data
        grid: null,
        comparisonGrid: null,
        probabilityGrid: null,
        minimumImprovementProbability: state.regional.minimumImprovementProbability,
        breaks: []
      }
    }
  },
  'set regional analysis bounds' (state, action) {
    return {
      ...state,
      regional: {
        bounds: action.payload
      }
    }
  },
  'set regional analysis grids' (state, action) {
    const { grid, comparisonGrid, probabilityGrid, differenceGrid, breaks } = action.payload
    return {
      ...state,
      regional: {
        ...state.regional,
        grid,
        comparisonGrid,
        probabilityGrid,
        differenceGrid,
        breaks
      }
    }
  },
  'set minimum improvement probability' (state, action) {
    return {
      ...state,
      regional: {
        ...state.regional,
        minimumImprovementProbability: action.payload
      }
    }
  },
  'set current indicator' (state, action) {
    return {
      ...state,
      currentIndicator: action.payload
    }
  },
  'select bookmark' (state, action) {
    const { profileRequest, isochroneCutoff, destinationGrid } = action.payload
    return {
      ...state,
      profileRequest,
      isochroneCutoff,
      isochroneLonLat: {
        lon: profileRequest.fromLon,
        lat: profileRequest.fromLat
      },
      currentIndicator: destinationGrid
    }
  }
}

export const initialState = {
  isochroneCutoff: 60,
  isFetchingIsochrone: false,
  active: false,
  scenarioApplicationErrors: null,
  regional: {
    id: null,
    comparisonId: null,
    percentile: 50,
    grid: null,
    comparisonGrid: null,
    probabilityGrid: null,
    minimumImprovementProbability: 0.95,
    breaks: []
  },
  profileRequest: {
    date: (new Date()).toISOString().split('T')[0],
    fromTime: 25200,
    toTime: 32400,
    accessModes: 'WALK',
    directModes: 'WALK',
    egressModes: 'WALK',
    transitModes: 'TRANSIT',
    walkSpeed: 1.3888888888888888,
    bikeSpeed: 4.166666666666667,
    carSpeed: 20,
    streetTime: 90,
    maxWalkTime: 20,
    maxBikeTime: 20,
    maxCarTime: 45,
    minBikeTime: 10,
    minCarTime: 10,
    suboptimalMinutes: 5,
    reachabilityThreshold: 0,
    bikeSafe: 1,
    bikeSlope: 1,
    bikeTime: 1,
    bikeTrafficStress: 4,
    monteCarloDraws: 200,
    maxRides: 4
  }
}
