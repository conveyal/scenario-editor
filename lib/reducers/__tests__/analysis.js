/* globals describe, expect, it */

import { handleActions } from 'redux-actions'

import * as analysis from '../analysis'
import {mockScenarioApplicationError} from '../../utils/mock-data'

describe('reducers > analysis', () => {
  const rawReducer = handleActions(analysis.reducers, analysis.initialState)
  // wrap reducer in order to snapshot test the ever-changing date
  const reducer = (state, action) => {
    const outState = rawReducer(state, action)
    if (outState && outState.profileRequest && outState.profileRequest.date) {
      outState.profileRequest.date = '2016-12-20'
    }
    return outState
  }

  // Default State Test
  it('should handle default state', () => {
    expect(reducer(undefined, { type: 'blah', payload: {} })).toMatchSnapshot()
  })

  // Specific Handler Tests
  it('should handle set profile request', () => {
    const action = {
      type: 'set profile request',
      payload: {
        maxWalkTime: 20,
        fromTime: 24321,
        toTime: 25220
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle clear isochrone results', () => {
    const action = { type: 'clear isochrone results', payload: {} }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set isochrone cutoff', () => {
    const action = { type: 'set isochrone cutoff', payload: 123 }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set isochrone fetch status', () => {
    const action = { type: 'set isochrone fetch status', payload: true }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set isochrone latlng', () => {
    const action = { type: 'set isochrone latlng', payload: 123 }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle show scenario application errors', () => {
    const action = { type: 'show scenario application errors', payload: [mockScenarioApplicationError] }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set isochrone results', () => {
    const action = {
      type: 'set isochrone results',
      payload: {
        isochrone: {},
        accessibility: 42,
        indicator: 'Jobs_total',
        grid: {},
        isochroneCutoff: 60,
        spectrogramData: ['SCENARIO'],
        comparisonAccessibility: 55,
        comparisonSpectrogramData: ['COMPARISON'],
        comparisonIsochrone: { comparison: true }
      }
    }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle enter analysis mode', () => {
    const action = { type: 'enter analysis mode' }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle exit analysis mode', () => {
    const action = { type: 'exit analysis mode' }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set active variant', () => {
    const action = {
      type: 'set active variant',
      payload: 21
    }
    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set comparison in progress', () => {
    const action = {
      type: 'set comparison in progress',
      payload: true
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set comparison scenario', () => {
    const action = {
      type: 'set comparison scenario',
      payload: {
        id: 'COMPARISON SCENARIO ID',
        bundleId: 'COMPARISON BUNDLE ID',
        variantIndex: 16
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set comparison modifications', () => {
    const action = {
      type: 'set comparison modifications',
      payload: {
        id: 'COMPARISON SCENARIO ID',
        bundleId: 'COMPARISON BUNDLE ID',
        variantIndex: 16,
        modifications: ['MODIFICATIONS']
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set active regional analyses', () => {
    const action = {
      type: 'set active regional analyses',
      payload: {
        id: 'scenario id',
        comparisonId: 'comparison id'
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })

  it('should handle set regional analysis grids', () => {
    const action = {
      type: 'set regional analysis grids',
      payload: {
        grid: { which: 'GRID' },
        comparisonGrid: { which: 'COMPARISON GRID' },
        differenceGrid: { which: 'DIFFERENCE_GRID' },
        probabilityGrid: { which: 'PROBABILITY GRID' }
      }
    }

    expect(reducer(undefined, action)).toMatchSnapshot()
  })
})
