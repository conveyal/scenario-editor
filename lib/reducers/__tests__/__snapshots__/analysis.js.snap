// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers > analysis should handle clear isochrone results 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochrone": null,
  "isochroneCutoff": 60,
  "isochroneLonLat": null,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle default state 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle enter analysis mode 1`] = `
Object {
  "active": true,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle exit analysis mode 1`] = `
Object {
  "active": false,
  "comparisonTravelTimeSurface": null,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
  "scenarioApplicationWarnings": null,
  "travelTimeSurface": null,
}
`;

exports[`reducers > analysis should handle set active regional analyses 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": "comparison id",
    "grid": null,
    "id": "scenario id",
    "minimumImprovementProbability": 0.95,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set active variant 1`] = `
Object {
  "active": false,
  "activeVariant": 21,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set comparison in progress 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set comparison modifications 1`] = `
Object {
  "active": false,
  "comparisonBundleId": "COMPARISON BUNDLE ID",
  "comparisonModifications": Array [
    "MODIFICATIONS",
  ],
  "comparisonScenarioId": "COMPARISON SCENARIO ID",
  "comparisonVariant": 16,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set comparison scenario 1`] = `
Object {
  "active": false,
  "comparisonBundleId": "COMPARISON BUNDLE ID",
  "comparisonModifications": null,
  "comparisonScenarioId": "COMPARISON SCENARIO ID",
  "comparisonVariant": 16,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set isochrone cutoff 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 123,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set isochrone fetch status 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": true,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set isochrone lonlat 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "isochroneLonLat": 123,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set profile request 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "fromTime": 24321,
    "maxWalkTime": 20,
    "toTime": 25220,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set regional analysis grids 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": undefined,
    "comparisonGrid": Object {
      "which": "COMPARISON GRID",
    },
    "comparisonId": null,
    "differenceGrid": Object {
      "which": "DIFFERENCE_GRID",
    },
    "grid": Object {
      "which": "GRID",
    },
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": Object {
      "which": "PROBABILITY GRID",
    },
  },
  "scenarioApplicationErrors": null,
}
`;

exports[`reducers > analysis should handle set scenario application errors 1`] = `
Object {
  "active": false,
  "isFetchingIsochrone": false,
  "isochroneCutoff": 60,
  "profileRequest": Object {
    "accessModes": "WALK",
    "bikeSafe": 1,
    "bikeSlope": 1,
    "bikeSpeed": 4.166666666666667,
    "bikeTime": 1,
    "bikeTrafficStress": 4,
    "carSpeed": 20,
    "date": "2016-12-20",
    "directModes": "WALK",
    "egressModes": "WALK",
    "fromTime": 25200,
    "maxBikeTime": 20,
    "maxCarTime": 45,
    "maxRides": 4,
    "maxWalkTime": 20,
    "minBikeTime": 10,
    "minCarTime": 10,
    "monteCarloDraws": 200,
    "reachabilityThreshold": 0,
    "streetTime": 90,
    "suboptimalMinutes": 5,
    "toTime": 32400,
    "transitModes": "TRANSIT",
    "walkSpeed": 1.3888888888888888,
  },
  "regional": Object {
    "breaks": Array [],
    "comparisonGrid": null,
    "comparisonId": null,
    "grid": null,
    "id": null,
    "minimumImprovementProbability": 0.95,
    "percentile": 50,
    "probabilityGrid": null,
  },
  "scenarioApplicationErrors": Array [
    Object {
      "messages": Array [
        "Headway is not greater than zero.",
      ],
      "modificationId": "Add Trip Pattern",
      "title": "Error while applying the modification entitled 'Add Trip Pattern'.",
    },
  ],
}
`;
